test/controllers/address/ChooseAddressControllerSpec.scala

```diff
@@ -0,0 +1,135 @@
+    super.beforeEach
+    reset(mockUserAnswersCacheConnector)
+  }
+
+  "Choose address controller" - {
+
+    "must return OK and the correct view for a GET" in {
+      val ua = emptyUserAnswers
+        .setOrException(EnterPostcodePage(Event1EmployerAddressJourney), seqTolerantAddresses)
+        .setOrException(CompanyDetailsPage, companyDetails)
+
+      val application = applicationBuilder(userAnswers = Some(ua)).build()
+
+      running(application) {
+        val request = FakeRequest(GET, getRoute)
+
+        val result = route(application, request).value
+
+        val view = application.injector.instanceOf[ChooseAddressView]
+
+        status(result) mustEqual OK
+        contentAsString(result) mustEqual
+          view(form, waypoints, Event1EmployerAddressJourney,
+            messages("chooseAddress.title", "the company"),
+            messages("chooseAddress.heading", companyDetails.companyName),
+            seqTolerantAddresses)(request, messages(application)).toString
+      }
+    }
+
+    "must save the answer and redirect to the next page when valid data is submitted" in {
+      val uaCaptor: ArgumentCaptor[UserAnswers] = ArgumentCaptor.forClass(classOf[UserAnswers])
+
+      when(mockUserAnswersCacheConnector.save(any(), any(), uaCaptor.capture())(any(), any()))
+        .thenReturn(Future.successful(()))
+
+      val ua = emptyUserAnswers.setOrException(EnterPostcodePage(Event1EmployerAddressJourney), seqTolerantAddresses)
+
+      val application =
+        applicationBuilder(userAnswers = Some(ua), extraModules)
+          .build()
+
+      running(application) {
+        val request =
+          FakeRequest(POST, postRoute).withFormUrlEncodedBody(("value", "0"))
+
+        val result = route(application, request).value
+        val updatedAnswers = emptyUserAnswers.set(ChooseAddressPage(Event1EmployerAddressJourney), seqAddresses.head).success.value
+
+        status(result) mustEqual SEE_OTHER
+        redirectLocation(result).value mustEqual ChooseAddressPage(Event1EmployerAddressJourney)
+          .navigate(waypoints, ua.setOrException(ChooseAddressPage(Event1EmployerAddressJourney), seqAddresses.head), updatedAnswers).url
+        verify(mockUserAnswersCacheConnector, times(1)).save(any(), any(), any())(any(), any())
+        uaCaptor.getValue.get(ManualAddressPage(Event1EmployerAddressJourney)) mustBe Some(seqAddresses.head)
+      }
+    }
+
+    "must return bad request when invalid data is submitted" in {
+      when(mockUserAnswersCacheConnector.save(any(), any(), any())(any(), any()))
+        .thenReturn(Future.successful(()))
+
+      val ua = emptyUserAnswers.setOrException(EnterPostcodePage(Event1EmployerAddressJourney), seqTolerantAddresses)
+
+      val application =
+        applicationBuilder(userAnswers = Some(ua), extraModules)
+          .build()
+
+      running(application) {
+        val request =
+          FakeRequest(POST, postRoute).withFormUrlEncodedBody(("value", "invalid"))
```

Comment:
shouldn't invalid-value be "" for these radio-button list ? and we may validate form and result ?

Reply:
Either would be invalid. I will change to empty string

Reply:
We don't really need to check result. So long as it is a BAD REQUEST we know it has failed validation. The form spec will test the specific error.